doc/Neighbor/Neighbor_Refresh_HLD.md

```diff
@@ -0,0 +1,321 @@
+  - On MAC delete (clear mac / control protocol) trigger refresh for corresponding ARP/ND
+
+## 1.2 Configuration and Management Requirements
+
+No New CLIs are added as part of this proposal.
+
+## 1.3 Scalability Requirements
+
+  - Support ARP/ND refresh for upto 4K Vlan Interfaces and 32K neighbor entries
+
+## 1.4 Warm Boot Requirements
+
+  - During Warmboot, MAC aging will be disabled and will be enabled back after the system is restared. Hence traffic blackholing due to MAC ageout will not happen. When the system boots up, ARP/ND refresh cache table will be populated again and then the neighbor entries will be refreshed as normal.
+
+## 1.5 Limitiations
+
+
+# 2 Functionality
+
+## 2.1 Functional Description
+
+Req-1:
+
+In SONiC, once IP address is configured on an interface, subnet route and IP2ME route entries are programmed. To forward traffic to any directly connected hosts on this subnet, packet will be first trapped to CPU based on the subnet route. Linux kernel on receiving the packet will trigger the neighbor (ARP/ND) resolution. Linux kernel handles the state machine for the neighbors resolved. Once the neighbor is resolved, NEIGH_TABLE entry will be created by neighsyncd in APP_DB, Orchagent will process the neighbor table entry and programs the SAI both the neighbor and the nexthop. Once the neighbor entry is programmed in HW/ASIC all further traffic to that destination will be forwarded by the ASIC and there is no notification/liveliness of the entry is known to Linux kernel. Based on the configured aging time, Linux kernel may change the state and could potentially remove the entry (depends on different kernel parameters). In order to keep the kernel neighbor state alive it is essential to refresh the neighbors by either send IP packets / ARP/NS packets to the destination host. Once the destination host responds kernel's state machine will get updated and all its neighbor timers are reset.
+
+Req-2:
+
+In case of Vlan interfaces, for an neighbor entry, in addition to the neighbor and nexthop entry programmed in ASIC, FDB entry corresponding to the resolved MAC to outgoing interface will be programmed. Without the FDB entry packets may not be forwarded in HW. Since the FDB resources are limited its quite normal to ageout an unsed FDB entry in HW. If the FDB entry ages out before the neighbor ageout then packets will be dropped in ASIC. So the neighbor refresh should ideally happen before the MAC/FDB ageout.
+
+Req-3:
+
+MAC table entry in SONiC can be administratively cleared or could be removed due to control protocol handling (mclag mac delete on orphan ports during mclag peer reboot). Once the MAC is removed by external triggers then traffic could be dropped up untill the next neighbor refresh which could be in minutes. In order to avoid such traffic drop, if a MAC is removed, neighbor entry should be refreshed as early as possible to avoid blackholing of traffic .
```

Comment:
When STP is enabled, there could be a possibility of seeing multiple topology changes before it settles down, how do we handle so many FDB flushes leading to adding same set of nbrs to the Q for refresh? do we hold on the refresh until the topology settles and then refresh all the nbrs or we refresh the MAC associated nbrs as and when the MAC is flushed in the HW? there is a CoPP limit for ARP/Nbr Rx in the HW, did we hit this limit in the testing i.e CoPP rule dropped the packets in the HW? if we dont receive the response, kernel takes are re-sending the packets but there  can be scenarios where user intervention is required to refresh the nbrs manually beyond the configured limit.